app/src/gui/components/authentication/auth_return.tsx

```diff
@@ -40,51 +39,100 @@ async function getListingForConductorUrl(conductor_url: string) {
   throw Error(`Unknown listing for conductor url ${conductor_url}`);
 }
 
-interface AuthReturnProps {
-  setToken: Dispatch<SetStateAction<TokenContents | null | undefined>>;
-}
-
-export function AuthReturn(props: AuthReturnProps) {
+export function AuthReturn() {
   const navigate = useNavigate();
+  const [error, setError] = useState<string | undefined>(undefined);
 
+  // track if effect has already run - this should only happen once
+  const hasRun = useRef(false);
+
+  const setErrorAndReturnHome = (msg: string) => {
+    console.error(msg);
+    setError(msg);
+    const timeout = setTimeout(() => {
+      navigate('/');
+    }, 10000);
+    return () => {
+      clearTimeout(timeout);
+    };
+  };
+
+  // This effect runs when the component mounts - it looks for a token in the query string
   useEffect(() => {
-    const storeToken = async (token: string) => {
-      const token_obj = decodeJwt(decodeURIComponent(token));
+    // don't run this effect twice as it causes a local pouch DB conflict
+    if (hasRun.current) return;
+    hasRun.current = true;
 
-      console.log('decoded', token_obj);
+    const storeToken = async (
+      token: string,
+      refreshToken: string | undefined
+    ) => {
+      // Decode in case URI encoded
+      const decodedToken = decodeURIComponent(token);
+      const decodedRefreshToken = refreshToken
+        ? decodeURIComponent(refreshToken)
+        : undefined;
 
-      const listing_id = await getListingForConductorUrl(
-        token_obj.server as string
-      );
+      // Decode the JWT object into an untyped object
+      const parsedToken = await parseToken(decodedToken);
+
+      // Get the listing for the server in the token
+      const listing_id = await getListingForConductorUrl(parsedToken.server);
+
+      // Store the token in the database
+      try {
+        await setTokenForCluster(
+          decodedToken,
+          parsedToken,
+          decodedRefreshToken,
+          listing_id
+        );
+      } catch (e) {
+        return setErrorAndReturnHome(
+          'Auth return route attempted to store token in local auth DB but encountered an error. ' +
```

Comment:
Seeing this error whenever I try to login: 
Error: Auth return route attempted to store token in local auth DB but encountered an error. TypeError: current_doc.available_tokens.get is not a function. You will be returned to home in 10 seconds.

Reply:
Sorry, it was the old db structure as you said.

Reply:
However, can we code around this at all? Would be a big black mark if users can't login after an update.

Reply:
This has been resolved as discussed by reverting to plain JS object but providing a wrapper ObjectMap class which provides a map interface over the object.

Reply:
In the future version tagging with migration strategy would be ideal.